File: Data Structures & Algorithms/combination-target-sum-ii/submission-7.java
class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        ans = new ArrayList<>();
        helper(candidates, 0, 0, target, new ArrayList<>());
        return ans;
    }
    public void helper(int[] nums, int i, int currsum, int t, List<Integer> curr) {
        if(currsum==t) {
            ans.add(new ArrayList<>(curr));
            return;
        }
        if(i==nums.length || currsum>t)
            return;
        curr.add(nums[i]);
        helper(nums,i+1,currsum+nums[i],t,curr);
        while(i<nums.length-1 && nums[i+1]==nums[i])
            i++;
        curr.removeLast();
        helper(nums,i+1,currsum,t,curr);
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-7.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans = []
        def helper(i: int, curr: List[int]):
            if i==len(nums) or sum(curr)>target:
                return
            if sum(curr)==target:
                ans.append(curr.copy())
                return
            curr.append(nums[i])
            helper(i, curr)
            curr.pop()
            helper(i+1, curr)
        helper(0,[])
        return ans